modality-probe-cli/src/manifest_gen/id_gen.rs

```diff
@@ -0,0 +1,207 @@
+use crate::exit_error;
+use core::num::NonZeroU32;
+use sha3::{Digest, Sha3_256};
+use std::hash::Hash;
+use std::str::FromStr;
+use uuid::Uuid;
+
+/// Inclusive non-zero ID (u32) range
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct NonZeroIdRange {
+    inclusive_start: NonZeroU32,
+    inclusive_end: NonZeroU32,
+}
+
+impl NonZeroIdRange {
+    pub fn new(inclusive_start: NonZeroU32, inclusive_end: NonZeroU32) -> Option<Self> {
+        if inclusive_start.get() > inclusive_end.get() {
+            None
+        } else {
+            Some(NonZeroIdRange {
+                inclusive_start,
+                inclusive_end,
+            })
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct IdGen {
+    id_range: NonZeroIdRange,
+    uuid: Uuid,
+}
+
+impl IdGen {
+    pub fn new(id_range: NonZeroIdRange) -> Self {
+        IdGen {
+            id_range,
+            uuid: Uuid::new_v4(),
+        }
+    }
+
+    pub fn regenerate_uuid(&mut self) {
+        self.uuid = Uuid::new_v4();
+    }
+
+    pub fn hashed_id(&mut self, token: &str) -> NonZeroU32 {
+        let mut max_tries = std::u16::MAX;
+        loop {
+            let hash = self.token_hash(token);
+            if let Some(non_zero_hash) = NonZeroU32::new(hash) {
+                if self.id_range.contains(non_zero_hash) {
+                    return non_zero_hash;
+                }
+            }
+
+            self.regenerate_uuid();
+
+            // Escape hatch
+            max_tries = max_tries.saturating_sub(1);
+            if max_tries == 0 {
+                exit_error!("manifest-gen", "Exceeded the id-hashing retry limit");
+            }
+        }
+    }
+
+    fn token_hash(&self, token: &str) -> u32 {
+        let mut hasher = Sha3_256::new();
+        hasher.update(self.uuid.as_bytes());
+        hasher.update(token.as_bytes());
+        let hash = hasher.finalize();
+        let bytes: &[u8; 32] = hash.as_ref();
+        let be16_bytes = u16::from_be_bytes([bytes[0], bytes[1]]);
+        let be32_bytes = u32::from_be_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]);
+        let id = u32::from(be16_bytes)
+            .overflowing_mul(0xFFFF_FFFF)
+            .0
+            .overflowing_add(be32_bytes)
+            .0;
+        id % self.id_range.inclusive_end.get()
+    }
+}
+
+impl NonZeroIdRange {
```

Comment:
Want to include this in the NonZeroIdRange impl block above?

Reply:
:+1: c2ba416587d310f0642db812259c0e3fd480c893